CS8620 warning on range indexers of strings

**Version Used**: 

Visual Studio 2019 Preview Version 16.0.0 Preview 1.1

**Steps to Reproduce**:

```cs
class Program
{
    static void Main()
    {
#nullable enable
        string s = "2134567890";
        string sub = s[1..3]; // CS8620
    }
}
```

**Expected Behavior**:

No waning.

**Actual Behavior**:

Warning	CS8620
Nullability of reference types in argument of type 'Range' doesn't match target type 'int' for parameter 'index' in 'char string.this[int index]'.
ConsoleApp1	C:\source\repos\ConsoleApp1\Program.cs	7	24

Thanks for reporting this. Confirmed.

@agocke Is there still some temporary logic in the compiler regarding indexers? 
Note that the diagnostic complains about `char string.this[int index]`, but I expect that a range-based indexer is used.

```C#
        [Fact, WorkItem(31748, "https://github.com/dotnet/roslyn/issues/31748")]
        public void RangeAndNullability()
        {
            var comp = CreateCompilationWithIndexAndRange(@"
#nullable enable
class C
{
    public static void M()
    {
        string s = ""2134567890"";
        string sub = s[1..3];
    }
}");
            comp.VerifyDiagnostics(
                // (8,24): warning CS8620: Nullability of reference types in argument of type 'Range' doesn't match target type 'int' for parameter 'index' in 'char string.this[int index]'.
                //         string sub = s[1..3];
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "1..3").WithArguments("System.Range", "int", "index", "char string.this[int index]").WithLocation(8, 24)
                );
        }
```

--- a/src/Compilers/CSharp/Test/Emit/CodeGen/IndexAndRangeTests.cs
+++ b/src/Compilers/CSharp/Test/Emit/CodeGen/IndexAndRangeTests.cs
@@ -111,7 +111,7 @@ class C
         MyString s = ""abcdef"";
         Console.WriteLine(s[^2..]);
     }
-}" + StringWithIndexers, expectedOutput: "ef");
+}" + TestSources.StringWithIndexers, expectedOutput: "ef");
         }
 
         [Fact]
@@ -126,7 +126,7 @@ class C
         MyString s = ""abcdef"";
         Console.WriteLine(s[^4..^1]);
     }
-}" + StringWithIndexers, expectedOutput: "cde");
+}" + TestSources.StringWithIndexers, expectedOutput: "cde");
         }
 
         [Fact]
@@ -146,7 +146,7 @@ class C
         Console.WriteLine(s[new Index(1, false)]);
         Console.WriteLine(s[new Index(1, false), ^1]);
     }
-}" + StringWithIndexers);
+}" + TestSources.StringWithIndexers);
             comp.VerifyDiagnostics(
                 // (13,27): error CS1501: No overload for method 'this' takes 2 arguments
                 //         Console.WriteLine(s[new Index(1, false), ^1]);
@@ -196,7 +196,7 @@ class C
         Console.WriteLine(s[new Index(1, false)]);
         Console.WriteLine(s[^1]);
     }
-}" + StringWithIndexers, expectedOutput: @"b
+}" + TestSources.StringWithIndexers, expectedOutput: @"b
 f");
             verifier.VerifyIL("C.M", @"
 {
@@ -232,7 +232,7 @@ class C
         Console.WriteLine(result);
     }
     public static string M(MyString s) => s[1..3];
-}" + StringWithIndexers, expectedOutput: "bc");
+}" + TestSources.StringWithIndexers, expectedOutput: "bc");
             verifier.VerifyIL("C.M", @"
 {
   // Code size       25 (0x19)
@@ -262,7 +262,7 @@ class C
         Console.WriteLine(result);
     }
     public static string M(MyString s) => s[1..];
-}" + StringWithIndexers, expectedOutput: "bcdef");
+}" + TestSources.StringWithIndexers, expectedOutput: "bcdef");
         }
 
         [Fact]
@@ -279,7 +279,7 @@ class C
         Console.WriteLine(result);
     }
     public static string M(MyString s) => s[..^2];
-}" + StringWithIndexers, expectedOutput: "abcd");
+}" + TestSources.StringWithIndexers, expectedOutput: "abcd");
         }
 
         [Fact]
@@ -1547,41 +1547,6 @@ partial class Program
         {
             return ""default"";
         }
-    }
-}";
-
-        // The references we use for System.String do not have an indexer for
-        // System.Index and System.Range, so this wrapper mimics the behavior
-        private const string StringWithIndexers = @"
-internal readonly struct MyString
-{
-    private readonly string _s;
-    public MyString(string s)
-    {
-        _s = s;
-    }
-    public static implicit operator MyString(string s) => new MyString(s);
-    public static implicit operator string(MyString m) => m._s;
-
-    public int Length => _s.Length;
-
-    public char this[int index] => _s[index];
-
-    public char this[Index index]
-    {
-        get
-        {
-            int actualIndex = index.GetOffset(Length);
-            return this[actualIndex];
-        }
-    }
-
-    public string this[Range range] => Substring(range);
-
-    public string Substring(Range range)
-    {
-        (int start, int length) = range.GetOffsetAndLength(Length);
-        return _s.Substring(start, length);
     }
 }";
     }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
@@ -507,6 +507,23 @@ public class MyClass
                 );
         }
 
+        [Fact, WorkItem(31748, "https://github.com/dotnet/roslyn/issues/31748")]
+        public void NullableRangeIndexer()
+        {
+            var text = @"
+using System;
+#nullable enable
+class Program
+{
+    void M(int[] x, MyString m)
+    {
+        var y = x[1..3];
+        var z = m[1..3];
+    }
+}" + TestSources.GetSubArray + TestSources.StringWithIndexers;
+            CreateCompilationWithIndexAndRange(text).VerifyDiagnostics();
+        }
+
         [Fact, WorkItem(31370, "https://github.com/dotnet/roslyn/issues/31370")]
         public void SuppressNullableWarning_MemberAccess()
         {

--- a/src/Compilers/Test/Utilities/CSharp/TestSources.cs
+++ b/src/Compilers/Test/Utilities/CSharp/TestSources.cs
@@ -385,5 +385,40 @@ namespace System.Runtime.CompilerServices
         }
     }
 }";
+
+        // The references we use for System.String do not have an indexer for
+        // System.Index and System.Range, so this wrapper mimics the behavior
+        public const string StringWithIndexers = @"
+internal readonly struct MyString
+{
+    private readonly string _s;
+    public MyString(string s)
+    {
+        _s = s;
+    }
+    public static implicit operator MyString(string s) => new MyString(s);
+    public static implicit operator string(MyString m) => m._s;
+
+    public int Length => _s.Length;
+
+    public char this[int index] => _s[index];
+
+    public char this[Index index]
+    {
+        get
+        {
+            int actualIndex = index.GetOffset(Length);
+            return this[actualIndex];
+        }
+    }
+
+    public string this[Range range] => Substring(range);
+
+    public string Substring(Range range)
+    {
+        (int start, int length) = range.GetOffsetAndLength(Length);
+        return _s.Substring(start, length);
+    }
+}";
     }
 }

